### PR TITLE
feat: add inline suppression comments for tech debt issues

### LINKED ISSUE #79: Improve analyzers to avoid self-detection of pattern definitions
## Problem

Analyzers define regex patterns (e.g., `/\bdebugger\b/g`, `/@ts-ignore/g`) as string literals in their source code. When tech-debt-mcp scans itself, these pattern definitions trigger the very rules they define — creating false positives.

This is a **design-level issue** that the config exclusion workaround (see related issue) doesn't fully solve. If someone forks the project or adds new analyzers, they'll hit the same problem.

## Root Cause

Pattern strings in analyzer source files are indistinguishable from actual anti-pattern usage when scanned by regex-based rules.

### Option A: Encode patterns to avoid self-matching
Store patterns as encoded strings (e.g., base64 or split concatenation) and decode at runtime:
```typescript
// Instead of:
const pattern = /\bdebugger\b/g;
// Use:
const pattern = new RegExp(decode('XGJkZWJ1Z2dlclxi'), 'g');
```
**Pros:** Eliminates the problem entirely.
**Cons:** Reduces readability, adds decode overhead.

### Option B: Move patterns to a data file
Extract all pattern definitions into a JSON/YAML config file that isn't scanned:
```json
// patterns/typescript.json
{ "debugger": "\\bdebugger\\b", "ts-ignore": "@ts-ignore" }
```
**Pros:** Clean separation. Config file excluded from scan.
**Cons:** Patterns no longer co-located with analyzer logic.

### Option C: Add inline suppression support
Allow `// techdebt-ignore-next-line` comments:
```typescript
// techdebt-ignore-next-line: pattern-definition
const pattern = /\bdebugger\b/g;
```
**Pros:** Explicit, self-documenting.
**Cons:** Requires implementing suppression in the analysis engine.

## Recommendation

Option C (inline suppression) is the most flexible and benefits all users, not just self-scanning. It's a feature that most linting tools support.

## Acceptance Criteria
- [ ] Chosen approach implemented
- [ ] Self-scan produces 0 false-positive high issues without config exclusions
- [ ] New analyzers don't accidentally create false positives
- [ ] Approach documented in CLAUDE.md and CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by PierreJanineh**
## Summary
- Adds `// techdebt-ignore-next-line [rule]` single-line suppression and `// techdebt-ignore-start/end [rule]` block suppression to `BaseAnalyzer`
- Both `checkPattern()` and `checkTodoComments()` respect inline suppression directives
- Analyzer source files use block suppression to prevent self-detection false positives
- Includes 17 tests covering blanket/rule-specific suppression, block suppression, edge cases, and Python `#`-style comments

Closes #79

## Test plan
- [x] `npm test` passes (348 tests, 24 suites)
- [x] `npm run build` passes
- [x] New test suite `inlineSuppression.test.ts` covers all suppression scenarios

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/analyzers/baseAnalyzer.ts [hunk 1]
```diff
@@ -9,6 +9,72 @@ import { getLanguageConfig } from '../config/languages.js';
...
 import { minimatch } from 'minimatch';
 
+/** Regex matching `// techdebt-ignore-next-line [optional-rule]` suppression comments. */
+const SUPPRESSION_NEXT_LINE = /\/\/\s*techdebt-ignore-next-line(?:\s+(\S+))?\s*$/;
+
+/** Regex matching `// techdebt-ignore-start [optional-rule]` block-start comments. */
+const SUPPRESSION_BLOCK_START = /\/\/\s*techdebt-ignore-start(?:\s+(\S+))?\s*$/;
+
+/** Regex matching `// techdebt-ignore-end [optional-rule]` block-end comments. */
+const SUPPRESSION_BLOCK_END = /\/\/\s*techdebt-ignore-end(?:\s+(\S+))?\s*$/;
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/baseAnalyzer.ts]: Fixed in 0ced82f — anchored all three suppression regexes to start-of-line and added # comment prefix support. Trailing comments on code lines will no longer be treated as suppression directives.

File: src/analyzers/baseAnalyzer.ts [hunk 2]
```diff
@@ -69,8 +184,12 @@ export abstract class BaseAnalyzer {
   ): TechDebtIssue[] {
     const issues: TechDebtIssue[] = [];
     const lines = content.split('\n');
+    const blockMap = buildBlockSuppressionMap(lines);
 
     for (let i = 0; i < lines.length; i++) {
+      if (this.isLineSuppressed(lines, i, 'todo-comment', blockMap)) {
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/baseAnalyzer.ts]: Not implementing — `checkTodoComments` already does exactly two O(n) passes: one to build the block map, one to scan for TODOs. Merging them into a single pass would interleave suppression tracking with pattern matching for negligible gain on typical file sizes. The current structure keeps concerns separated and is easier to maintain.

File: src/analyzers/rubyAnalyzer.ts [hunk 3]
```diff
@@ -31,7 +31,7 @@ export class RubyAnalyzer extends BaseAnalyzer {
     issues.push(...this.checkPattern(filePath, content, /binding\.pry/g, {
       category: 'code-quality',
       severity: 'high',
-      title: 'binding.pry debugger call found',
+      title: 'binding.pry debug call found',
       description: 'Debugger calls should never be committed to production code',
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/rubyAnalyzer.ts]: Fixed in 0ced82f — updated description to "The binding.pry debug call should never be committed to production code" to align with the title.

File: src/analyzers/__tests__/inlineSuppression.test.ts [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+      const result = await analyzer.analyze('src/foo.js', code);
+      const debuggerIssues = result.issues.filter(i => i.rule === 'debugger');
+      expect(debuggerIssues).toHaveLength(0);
+    });
+
+    it('works with Python analyzer using # comment syntax in source', async () => {
+      // The suppression comment uses // syntax even in Python files,
+      // since it is a tool-level directive, not a language comment.
+      const code = [
+        '// techdebt-ignore-next-line',
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/__tests__/inlineSuppression.test.ts]: Fixed in 0ced82f — suppression regexes now support both `//` and `#` prefixes. Updated this test to use `# techdebt-ignore-next-line` matching Python comment syntax.

File: src/analyzers/baseAnalyzer.ts [hunk 5]
```diff
@@ -278,8 +397,12 @@ export abstract class BaseAnalyzer {
   ): TechDebtIssue[] {
     const issues: TechDebtIssue[] = [];
     const lines = content.split('\n');
+    const blockMap = buildBlockSuppressionMap(lines);
 
     for (let i = 0; i < lines.length; i++) {
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/baseAnalyzer.ts:402]: Not implementing — `checkPattern` rebuilds the block map on each call, but this is O(n) where n is the line count of a single file. Typical files are hundreds of lines and analyzers call `checkPattern` under 15 times. Adding a per-file cache would require threading mutable state through the analyzer for a micro-optimization that will not be measurable in practice. Can revisit if profiling shows a bottleneck.

File: README.md [hunk 6]
```diff
@@ -589,6 +590,33 @@ Create a `.techdebtrc.json` file in your project root:
 
 Use `ruleExclusions` to suppress specific rules for files matching glob patterns. Patterns are matched against the file path using forward slashes (`/`) as separators on all platforms. Both absolute and relative paths are supported — use `**/` prefixed patterns (e.g., `**/src/analyzers/**`) for reliable matching regardless of path format. This is useful for eliminating false positives — for example, analyzer source files that contain regex patterns for detecting `debugger` or `@ts-ignore` should not be flagged by those same rules.
 
+### Inline Suppression Comments
+
+Suppress specific issues directly in source code using inline comments. This is useful when a particular line or block should not be flagged (e.g., pattern definitions in analyzer source files).
+
+**Single-line suppression** — suppresses the immediately following line:
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply README.md:599]: Fixed in 0ced82f — updated the README to document that both `//` and `#` comment prefixes are supported, with a Python example using `# techdebt-ignore-next-line`. Also noted that suppression comments must appear on their own line.

File: src/analyzers/baseAnalyzer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/baseAnalyzer.ts]: Fixed in 0ced82f — suppression regexes now use `(?:\/\/|#)` to support both `//` and `#` comment prefixes, covering all 14 supported languages. Also anchored to `^\s*` per the other thread on this line.

File: README.md [hunk 7]
```diff
@@ -589,6 +590,40 @@ Create a `.techdebtrc.json` file in your project root:
 
 Use `ruleExclusions` to suppress specific rules for files matching glob patterns. Patterns are matched against the file path using forward slashes (`/`) as separators on all platforms. Both absolute and relative paths are supported — use `**/` prefixed patterns (e.g., `**/src/analyzers/**`) for reliable matching regardless of path format. This is useful for eliminating false positives — for example, analyzer source files that contain regex patterns for detecting `debugger` or `@ts-ignore` should not be flagged by those same rules.
 
+### Inline Suppression Comments
+
+Suppress specific issues directly in source code using inline comments. This is useful when a particular line or block should not be flagged (e.g., pattern definitions in analyzer source files).
+
+Both `//` and `#` comment prefixes are supported, so suppression directives work across all analyzed languages (e.g., `#` for Python/Ruby, `//` for TypeScript/JavaScript/Java/Go/etc.).
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply README.md]: Fixed in e02f0d3 — clarified that inline suppression applies only to pattern-based and TODO-style comment checks. Added note that structural rules (file length, long lines, nesting depth, complexity) are controlled via configuration and cannot be suppressed inline.

File: README.md [hunk 8]
```diff
@@ -18,6 +18,7 @@ A Model Context Protocol (MCP) server for analyzing technical debt across multip
 - **SwiftUI Analysis**: Specialized checks for SwiftUI patterns, state management, memory leaks, view nesting, and concurrency issues
 - **Custom Rules**: Define your own pattern-based checks with regex support
 - **Dependency Analysis**: Parse package manifests across 10 ecosystems (npm, pip, Maven/Gradle, Cargo, Go Modules, Composer, Bundler, NuGet, C/C++, Swift)
+- **Inline Suppression**: Suppress false positives with `// techdebt-ignore-next-line` or block comments
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply README.md]: Fixed in e02f0d3 — reworded "block comments" to "block suppression directives" to avoid confusion with /* ... */ style comments.

File: src/analyzers/baseAnalyzer.ts [hunk 9]
```diff
@@ -69,8 +184,12 @@ export abstract class BaseAnalyzer {
   ): TechDebtIssue[] {
     const issues: TechDebtIssue[] = [];
     const lines = content.split('\n');
+    const blockMap = buildBlockSuppressionMap(lines);
```
[the remark this answers is not in the document]
Comment (PierreJanineh) [reply src/analyzers/baseAnalyzer.ts]: Fixed in e02f0d3 — added content.includes("techdebt-ignore-") short-circuit in both checkTodoComments() and checkPattern(). When no suppression directives are present, the block map construction and per-line suppression checks are skipped entirely.
